### ISSUE TITLE
beta regression matching unit-like structs with ..

### ISSUE THREAD

**Issue by durka** [rust-lang/rust#30379]
This code

```rust
struct Empty;

match Empty {
    Empty(..) => {}
}
```

compiles on stable 1.5 (no RFC 218 warning) but fails on beta 1.6 with E0164.

**Comment by petrochenkov** [comment]
This is an expected result of https://github.com/rust-lang/rust/pull/29383

**Comment by durka** [comment]
@petrochenkov breaking change with no warning cycle? I expected a warning like there is with a unitary enum variant.

**Comment by petrochenkov** [comment]
The warning was introduced only for the case `UnitVariant(..)`, other cases were not encountered during crater run.

**Comment by durka** [comment]
I guess my crate wasn't up during the crater run then :( But this is the point of beta, right, to catch stuff like this?

**Comment by petrochenkov** [comment]
I don't know, it's possible to turn all errors produced by https://github.com/rust-lang/rust/pull/29383 into warnings temporarily, but IMHO it's simpler to fix the bug on your side.
This is a policy question, so let's someone else decide.
cc @nikomatsakis  as a reviewer of https://github.com/rust-lang/rust/pull/29383

**Comment by alexcrichton** [comment]
triage: I-nominated

**Comment by durka** [comment]
In this case it's in a complex macro (my [guard](https://crates.io/crates/guard/) crate) so there really is no workaround until `#![feature(braced_empty_structs)]`. But you may overrule me by saying tt-based macros are never stable :)

**Comment by petrochenkov** [comment]
@durka 
>In this case it's in a complex macro (my guard crate)

Hmm, does something like `self::Empty` work as a workaround for this enum/identifier problem?

**Comment by durka** [comment]
Yes (though it's ugly), unless the struct/enum is local to the function/block. So it's a rather niche case I suppose.

**Comment by durka** [comment]
@petrochenkov actually no, that doesn't seem to help for structs imported from other modules: http://is.gd/8IcOM9

**Comment by petrochenkov** [comment]
@durka 
Then the module name can be used instead of `self` - `foo::Empty`.
I know this is all horrible, but `Empty(..)` itself is a workaround too.

**Comment by pnkfelix** [comment]
Hmm, I must have misunderstood #29383 ; I had thought there *was* a warning cycle associated with this change... e.g. wasn't this [comment](https://github.com/rust-lang/rust/pull/29383#issuecomment-160652130) evidence of such a warning?  What did I miss here...

**Comment by durka** [comment]
@pnkfelix in beta, enums warn while structs error.

**Comment by pnkfelix** [comment]
> @pnkfelix in beta, enums warn while structs error.

Ah!  Okay, I see.

**Comment by pnkfelix** [comment]
@petrochenkov 

> it's possible to turn all errors produced by #29383 into warnings temporarily, but IMHO it's simpler to fix the bug on your side. This is a policy question, so let's someone else decide.

(I think the appropriate action here is to make them all warnings.  But hopefully we'll settle this at the compiler team meeting in a few days.)

**Comment by petrochenkov** [comment]
@durka 
By the way, inability to use identifier in that position is not a parsing/expansion problem, it's a name resolution quirk. Patterns in `match` and `let` behave differently - `let` unlike `match` never considers identifiers to be variants or consts with rationale ["it's rarely what you want"](https://github.com/rust-lang/rust/blob/master/src/librustc_resolve/lib.rs#L2519)

```
struct S;

fn main() {
    let S = S; // error: declaration of `S` shadows an enum variant or unit-like struct in scope
}
```

**Comment by durka** [comment]
Oh interesting, I didn't realize that was a hard error in some places.

**Comment by petrochenkov** [comment]
I still think reverting the error doesn't worth the effort.
The broken code in https://github.com/durka/guard is a test code, testing some pretty contrived construction equivalent to `if let S(..) = S { /*do nothing*/ } else { return }` or `if false { return }`.

**Comment by nikomatsakis** [comment]
On Mon, Dec 14, 2015 at 03:52:40PM -0800, Vadim Petrochenkov wrote:
> By the way, inability to use identifier in that position is not a parsing/expansion problem, it's a name resolution quirk. Patterns in `match` and `let` behave differently - `let` unlike `match` never considers identifiers to be variants or consts with rationale ["it's rarely what you want"](https://github.com/rust-lang/rust/blob/master/src/librustc_resolve/lib.rs#L2519)

Huh, yet an other quirk of resolve. I can't say I agree with that
logic. Well, it IS rarely what you want, but I'd probably still treat
them the same! Sigh.

**Comment by pnkfelix** [comment]
> it's possible to turn all errors produced by #29383 into warnings temporarily, but [...]

Compiler team discussed this and decided that we'd prefer to take this tack.

I'll take care of this (after all, I overlooked it in my review of the original PR).

**Comment by brson** [comment]
@pnkfelix Two weeks until this is released. Needs a fix soon.

**Comment by pnkfelix** [comment]
@brson thanks, lost this in shuffle of other work

**Comment by pnkfelix** [comment]
triage: P-high

**Comment by pnkfelix** [comment]
This is a beta issue; it was an accident for me to close it in PR #30753 itself, since that solely affected nightly.

**Comment by brson** [comment]
Merged on beta.
